README.md
# Diffconvpl
Version: 0.1.3

##
As of version 0.2.0 this project requires Python 3.10.* .
***
## General
***



### Installation of requirements
In order to be able to run the source code the required packages need to be installed.
This can be done via pip with the following command:

`pip install -r requirements_dev.txt`

In order to run the tests please install the required packages via

`pip install -r requirements_test.txt`

### Running the source code

Locate the `application.py` and simply run it with Python. This will run with the following default arguments

        APP_ARGS = {
            "func": "torch.sin(10*x)",
            "params": ["x"],
            "m": 32,
            "entries": 100000,
            "epochs": 5000,
            "positive_funcs": 16,
            "negative_funcs": 16,
            "fullplot": fullplot,
        }


The source code will now try to build a model that approximates the given `func`. Parameters can be changed to approximate other functions as well.

### Commandline Arguments

The program accepts certain commandline Arguments:

`--autosave` will automatically save the generated output data without user-intercation.

`--fullplot` will add all evaluated MaxAffineFunctions to the output plot.

`--no-batch` will disable batched computation.

`--autorun <str:FILEPATH>` will recreate the model with parameters provided by a given filepath. Use the files within `data\json`.

`--batchsize <int:AMOUNT>` will set the size of batches for computation (only use, when large datasets are given)

## SDF Generators
***

The branch `sdf2d` contains scripts that generate `JSON`-Files that can be applied to the `application.py` using the `--autorun` command line argument.
Located under **`src\generators`** the scripts will produce the aforementioned `JSON`-Files under **`data\generated`** .

### Example call

`python .\application.py --autorun data\generated\2DSDF_Rectangle_BETA.json --no-batch --autosave`

### Generators and Commandline Arguments

The following generators exist and provide the functionality to parse specific commandline arguments as shown below:

*Note: The square generator can only handle datapoints in increments of 4.*
*Note: The triangle generator can only handle datapoints in increments of 3.*



## Findings

With the current setup ( as of 2023/01/12 ) the model cannot approximate an SDF properly when feeded "perfect" input data.
A small randomized offset suffices in order properly approximate the given SDf from data.

Observed via:

`python .\src\generators\sphere.py --setting distance --radius 1.0 --datapoints 100 --delta 0.0` --> Leads to the inability to approximate.

`python .\src\generators\sphere.py --setting distance --radius 1.0 --datapoints 100 --delta 0.1` --> Proper approximation possible.

Both scenarios get called via:

`python .\application.py --autorun .\data\generated\2DSDF_Circle.json --no-batch --autosave`


changelog.txt
Version 0.2.2
    Findings:
        -   due to the differences in the approach between both models (ReLu-MLP and Max-Affine) it is hard to define a proper mapping
            between the different parameters such that both models have eqal requirements in order to calculate a model
        -   in addition to the point above, aiming for a configuration which only compares the calculation and repsectively the result itself,
            when given identical inputs, has to be achieved
    Changes:
        -   a first (yet) untested implementation of a ReLu-MLP has been done and will be used as comparison
        -   a setup script has been written, that automates the setup of a virtual enviroment (it was planned to be added in version 0.2.4)
        -   polling datapoints from the resulting model is now vectorized (from current tests the speed has increased by ~250x)
    Future Changes:
        -   reintegration and redesign of the error propagation will be implemented in version 0.2.3
        -   the new setup script and the implementation of a ReLu-MLP will lead to a general refactorization

Version 0.2.1
    Findings:
        -   the generated sdf seems to be stable only in the range of provided data (distortions are visible below and above)
        -   polling datapoints from the resulting model is reasonably fast, but becomes very slow with increasing m and the amount of functions
            a vectorized approach has to be found to increase the speed of computation
    Changes:
        -   added BlenderSceneHandler.layercut
        -   old data generators under `src\generators` have been deprecated and removed
        -   sdf_generator will replace the removed generators
        -   pyexcel and code that is using it have been removed
        -   due to the changes above there is no implemented way to gain knowledge about the error propagation (therefore, the appropriate sequences
            in the code have been deactivated though commenting)
    Future Changes:
        -   a ReLu-MLP will be added in version 0.2.2 and used as comarison
        -   reintegration and redesign of the error propagation will be implemented in version 0.2.3
        -   an automated setup will be added until version 0.2.4

Version 0.2.0
    Findings:
        -   introducting a wider spread of data (i.e. more noise) increases the range of correct/fitting approximations significantly
    Changes:
        -   version-fixing in requirements_dev text file
        -   used Python version changed (3.9* --> 3.10.*) to accomodate bpy (the Blender Python module)
        -   added MemorySharedSubprocess to avoid errorless termination when passing data to the plot mechanism
        -   a lot of unused code is removed
    Future Changes:
        -   data generators in `src\generators` will be removed in version 0.2.1 and replaced with BlenderSceneHandler.layercut

Version 0.1.2
    Findings:
        -   currently the algorithm is not able to approximate an SDF when perfect input data is given (i.e. all datapoints have a distance of ~0.0 from the surface)
    Changes:
        -   Added the a "grid" setting to SDF data generators (currently only for the circle generator)
        -   Code optimizations and cleanup
        -   The application does now use a contour plot in order to visualize the results. This will later be extended with a heightmap like plot in 3D.

Version 0.1.1
    Changes:
        -   Added data SDF data generators (square,triangle,circle).


Version 0.1
    This is a port from main branch.
    Changes:
        -   Added the ability to work with 2-dimensional input data.


application.py
import argparse
import os
import random
from datetime import datetime
from typing import Any, Optional

import numpy as np
import torch
from tqdm import tqdm

from src.models.maxaffine import MultiDimMaxAffineFunction
from src.util.common import *
from src.util.logger import ParamLogger
from src.util.parameter_initializer import Initializer
from src.util.plot import plotsdf
from src.util.shm_process import MemorySharedSubprocess

random.seed(os.environ.get("GLOBAL_SEED"))

AUTOSAVE: bool = False
APP_ARGS: dict[str, Any] = {}


def run():
    global APP_ARGS, AUTOSAVE
    print("STAGE: Setup")
    fullplot = APP_ARGS["fullplot"]
    logger = ParamLogger()

    m = APP_ARGS["m"]
    entries = APP_ARGS["entries"]
    epochs = APP_ARGS["epochs"]
    positive_funcs = APP_ARGS["positive_funcs"]
    negative_funcs = APP_ARGS["negative_funcs"]
    display = APP_ARGS["display_blender"]

    batching = "batch_size" in APP_ARGS.keys()
    if batching:
        batch_size = APP_ARGS["batch_size"]

    signs: np.array = np.asarray(
        make_signs(positive=positive_funcs, negative=negative_funcs)
    )
    k: int = len(signs)

    datapoints: np.array = np.asanyarray([(d[0], d[1]) for d in APP_ARGS["data"]])

    y: torch.Tensor = (
        torch.from_numpy(np.asarray([dp[2] for dp in APP_ARGS["data"]]))
        .type(torch.FloatTensor)
        .to(torch.device("cuda:0"))
    )

    model: MultiDimMaxAffineFunction = MultiDimMaxAffineFunction(
        m=m,
        k=k,
        dim=2,
        x=datapoints,
        y=y,
        signs=signs,
        initializer=Initializer("src\\initializations\\parabola.json", dim=2),
        batchsize=batch_size if batching else 2**32,
    ).to(torch.device("cuda:0"))

    if batching == True:
        print("STAGE: Calculating batchsize")
        print_colored(
            "Warning: Running computation in batched mode is significantly slower. Please check whether it is needed!",
            cmdcolors.WARNING,
        )
        if APP_ARGS["batch_size"] == 2**32:
            model.bench(y_target=y, granularity=0.01)
            print(f"       Maximum applicaple batchsize={model.batch_size}")
        else:
            print(f"       Using preset batchsize of {batch_size}")
            model.batches = get_batch_spacing(batch_size, model.x.shape[0])

    print("STAGE: Calculation")

    loss_plot: list[float] = []
    pbar = tqdm(total=epochs)
    for _ in range(epochs):
        model.forward(batching)
        loss_plot.append(model.get_loss())
        pbar.update(1)

    pbar.close()
    pbar_dict = pbar.format_dict

    print("STAGE: Model Evaluation")

    timetag = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
    with open(os.path.join(os.getcwd(), "current_timetag.txt"), "w") as fp:
        fp.write(timetag)
        fp.close()

    spacing = 0.01
    points: np.array = int((abs(-1.0) + abs(1.0)) / spacing) + 1
    domain: np.array = np.linspace(-1.0, 1.0, points)
    x, y = np.meshgrid(domain, domain)

    z: np.ndarray = np.zeros_like(x)
    for ki in tqdm(range(model.k)):
        z += model.generate_sdf_plot_data(x=x, y=y, k=ki)

    # try:
    #     error_domain, error_values = model.error_propagation(
    #         spacing=0.01, min=-20.0, max=20.0
    #     )
    # except:
    #     print("ERROR while generating error propagation")

    plot_dict = {
        "func": "Unkown",
        "xv": x.tolist(),
        "yv": y.tolist(),
        "z": z.tolist(),
        "err_d": [],
        "err_v": [],
        "autosave": AUTOSAVE,
        "losses": loss_plot,
        "filename": timetag,
        "display_blend": display,
    }

    with open(
        os.path.join(
            os.getcwd(),
            "data",
            "generated",
            "blender_files",
            f"datapoints_{timetag}.txt",
        ),
        "w",
    ) as dp_file:
        for dp in APP_ARGS["data"]:
            dp_file.write(f"{dp[0]:.4f} {dp[1]:.4f} {dp[2]:.4f}\n")

    shared_memory_process = MemorySharedSubprocess(target=plotsdf)
    shared_memory_process.fork_on(data=plot_dict)
    shared_memory_process.await_join()

    print("STAGE: End")

    keep_data: Optional[bool]
    success: Optional[bool]
    if AUTOSAVE == True:
        keep_data = True
        success = True
    else:
        keep_data = None
        success = None

    while keep_data is None:
        answer = input("Do you want to keep the used parameters? (y/n)\n")
        if answer in ["y", "n"]:
            keep_data = "y" == answer
        while success is None:
            answer = input("Is the fitting successful? (y/n)\n")
            if answer in ["y", "n"]:
                success = "y" == answer
                break

    if keep_data:
        log_dict = build_log_dict(
            tqdm_dict=pbar_dict,
            loss=loss_plot[-1],
            func="Unknown",
            positive=positive_funcs,
            negative=negative_funcs,
            success=success,
            autosave=AUTOSAVE,
        )
        logger.full_log(dict=log_dict)
        APP_ARGS["Success"] = success
        APP_ARGS["Autosave"] = AUTOSAVE
        logger.json_log(dict=APP_ARGS, filename=timetag)

    if os.environ.get("diffconvpl_running") is not None:
        os.environ["diffconvpl_running"] = "0"


def setup(
    autosave: bool,
    fullplot: bool,
    filepath: Optional[str] = None,
    batch_size: int = 2**32,
    no_batch: bool = False,
    display_blender: bool = False,
) -> None:
    global APP_ARGS, AUTOSAVE
    AUTOSAVE = autosave
    if not filepath:
        for directory in ["data", "data\\json", "data\\plots", "data\\generated"]:
            if not os.path.exists(directory):
                os.makedirs(directory)

        APP_ARGS = {
            "func": "torch.sin(10*x)",
            "params": ["x"],
            "m": 32,
            "entries": 100000,
            "epochs": 5000,
            "positive_funcs": 16,
            "negative_funcs": 16,
            "fullplot": fullplot,
        }

        if no_batch == False:
            APP_ARGS["batch_size"] = batch_size

    else:
        APP_ARGS = rerun_experiment(filepath)
    APP_ARGS["display_blender"] = display_blender


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Diffconvpl")

    parser.add_argument(
        "--autosave",
        action="store_true",
        help="Skips questions and directly saves data.",
    )

    parser.add_argument(
        "--fullplot",
        action="store_true",
        help="Enable plotting all max-affine functions.",
    )

    parser.add_argument(
        "--no-batch", action="store_true", help="Disables batch-computation."
    )

    parser.add_argument(
        "--autorun",
        metavar="str",
        default=None,
        action="store",
        nargs="?",
        help="Issues and autorun with a given configuration via file.",
    )

    parser.add_argument(
        "--batchsize",
        metavar="int",
        default=2**32,
        action="store",
        nargs="?",
        help="Presets batch_size value used for large datasets",
    )

    parser.add_argument(
        "--blender-render",
        metavar="bool",
        default=False,
        action="store",
        nargs="?",
        help="Wether to run blender and display the results after computation.",
    )
    args = parser.parse_args()

    setup(
        autosave=args.autosave,
        fullplot=args.fullplot,
        filepath=args.autorun,
        batch_size=int(args.batchsize),
        no_batch=args.no_batch,
        display_blender=args.blender_render,
    )
    run()
